src/crypto/ecies.ts

```diff
@@ -0,0 +1,265 @@
+  assert(isValidPrivateKey(privateKey), 'Bad private key')
+  // See https://github.com/wanderer/secp256k1-node/issues/46
+  const compressed = true
+  return Buffer.from(
+    ec.keyFromPrivate(privateKey).getPublic(compressed, 'array')
+  )
+}
+
+// NOTE(Kagami): We don't use promise shim in Browser implementation
+// because it's supported natively in new browsers (see
+// <http://caniuse.com/#feat=promises>) and we can use only new browsers
+// because of the WebCryptoAPI (see
+// <http://caniuse.com/#feat=cryptography>).
+export function sign(privateKey: Buffer, msg: Buffer) {
+  return new Promise(function (resolve) {
+    assert(privateKey.length === 32, 'Bad private key')
+    assert(isValidPrivateKey(privateKey), 'Bad private key')
+    assert(msg.length > 0, 'Message should not be empty')
+    assert(msg.length <= 32, 'Message is too long')
+    resolve(Buffer.from(ec.sign(msg, privateKey, { canonical: true }).toDER()))
+  })
+}
+
+export function verify(publicKey: Buffer, msg: Buffer, sig: Buffer) {
+  return new Promise(function (resolve, reject) {
+    assert(publicKey.length === 65 || publicKey.length === 33, 'Bad public key')
+    if (publicKey.length === 65) {
+      assert(publicKey[0] === 4, 'Bad public key')
+    }
+    if (publicKey.length === 33) {
+      assert(publicKey[0] === 2 || publicKey[0] === 3, 'Bad public key')
+    }
+    assert(msg.length > 0, 'Message should not be empty')
+    assert(msg.length <= 32, 'Message is too long')
+    if (ec.verify(msg, sig, publicKey)) {
+      resolve(null)
+    } else {
+      reject(new Error('Bad signature'))
+    }
+  })
+}
+
+export function derive(
+  privateKeyA: Buffer,
+  publicKeyB: Buffer
+): Promise<Buffer> {
+  return new Promise(function (resolve) {
+    assert(Buffer.isBuffer(privateKeyA), 'Bad private key')
+    assert(Buffer.isBuffer(publicKeyB), 'Bad public key')
+    assert(privateKeyA.length === 32, 'Bad private key')
+    assert(isValidPrivateKey(privateKeyA), 'Bad private key')
+    assert(
+      publicKeyB.length === 65 || publicKeyB.length === 33,
+      'Bad public key'
+    )
+    if (publicKeyB.length === 65) {
+      assert(publicKeyB[0] === 4, 'Bad public key')
+    }
+    if (publicKeyB.length === 33) {
+      assert(publicKeyB[0] === 2 || publicKeyB[0] === 3, 'Bad public key')
+    }
+    const keyA = ec.keyFromPrivate(privateKeyA)
+    const keyB = ec.keyFromPublic(publicKeyB)
+    const Px = keyA.derive(keyB.getPublic()) // BN instance
+    resolve(Buffer.from(Px.toArray()))
+  })
+}
+
+export async function encrypt(
+  publicKeyTo: Buffer,
+  msg: Buffer,
+  opts?: { ephemPrivateKey?: Buffer; iv?: Buffer } | undefined
+) {
+  opts = opts || {}
+  // Take IV from opts or generate randomly
+  const iv = opts?.iv || randomBytes(16)
+  let ephemPrivateKey = opts?.ephemPrivateKey || randomBytes(32)
+  // There is a very unlikely possibility that it is not a valid key
+  while (!isValidPrivateKey(ephemPrivateKey)) {
+    if (opts?.ephemPrivateKey) {
```

Comment:
I added this if statement and error throwing. This was a bug in `eccrypto`. If an invalid `opts.ephemPrivateKey` is provided, the while loop would run infinitely in [their implementation](https://github.com/bitchan/eccrypto/blob/master/browser.js#L216).

Reply:
Wow nice find. Kind of crazy to see that while loop just repeatedly reassigning `ephemPrivateKey = opts?.ephemPrivateKey` in the case it's specified.